File: 백준/Gold/10026. 적록색약/적록색약.java
import java.io.*;

public class Main {
    static char[][] graph;
    static boolean[][] visited;
    static boolean[][] colorVisited;
    static int[] xDirection = new int[] {1, -1, 0, 0};
    static int[] yDirection = new int[] {0, 0, 1, -1};


    static void dfs(int r, int c, boolean check ) {
        boolean[][] visit;
        if (check) {
            visit = colorVisited;
        } else {
            visit = visited;
        }
        visit[r][c] = true;

        for (int i = 0; i < 4; i++) {
            int nx = r + xDirection[i];
            int ny = c + yDirection[i];

            if (nx < 0 || nx >= visited.length || ny < 0 || ny >= visited.length) {
                continue;
            } else if (visit[nx][ny]) {
                continue;
            } else if (graph[r][c] == graph[nx][ny]) {
                dfs(nx, ny, check);
            }

            if (check && graph[nx][ny] != 'B' && graph[r][c] != 'B') {
                dfs(nx, ny, check);
            }
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        int normal = 0;
        int color = 0;
        visited = new boolean[n][n];
        colorVisited = new boolean[n][n];
        graph = new char[n][n];

        for (int i = 0; i < n; i++) {
            String row = br.readLine();
            for (int j = 0; j < n; j++) {
                graph[i][j] = row.charAt(j);
            }
        }

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (!visited[i][j]) {
                    dfs(i, j, false);
                    normal += 1;
                } 
                
                if (!colorVisited[i][j]) {
                    dfs(i, j, true);
                    color += 1;
                }
            }
        }
        System.out.println(normal + " " + color);
    }
}

File: 백준/Gold/10026. 적록색약/적록색약.py
# 적록색약
import sys
input = sys.stdin.readline
sys.setrecursionlimit(100001)

def dfs(r, c, check):
    if check:
        visit = colorVisited
    else:
        visit = visited
        
    visit[r][c] = 1
    direction = [(1, 0), (-1, 0), (0, 1), (0, -1)]

    for x, y in direction:
        nx, ny = r + x, c + y
        if nx >= n or nx < 0 or ny >= n or ny < 0:
            continue
        elif visit[nx][ny]:
            continue
        elif graph[r][c] == graph[nx][ny]:
            dfs(nx, ny, check)

        # 색약인경우    
        if check:
            if (graph[r][c] == "R" or graph[r][c] == "G") and (graph[nx][ny] == "R" or graph[nx][ny] == "G"):
                dfs(nx, ny, check)
        
n = int(input())
graph = []
normal = 0 # 일반
color = 0 # 적록색약
visited = [[0] * n for _ in range(n)]
colorVisited = [[0] * n for _ in range(n)]

for i in range(n):
    graph.append(input().rstrip())

for i in range(n):
    for j in range(n):
        if not visited[i][j]:
            dfs(i, j, 0)
            normal += 1
        if not colorVisited[i][j]:
            dfs(i, j, 1)
            color += 1

print(normal, color)
